[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU

howardzhang-cv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171160
* __->__ #170486

Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Chillee @drisspg @yanboliang @BoyuanFeng

Differential Revision: [D89899238](https://our.internmc.facebook.com/intern/diff/D89899238)

howardzhang-cv committed:
[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU
Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

[ghstack-poisoned]

howardzhang-cv committed:
Update on "[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU"


Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

howardzhang-cv committed:
Update

[ghstack-poisoned]

howardzhang-cv committed:
Update

[ghstack-poisoned]

howardzhang-cv committed:
Update

[ghstack-poisoned]

howardzhang-cv committed:
Update

[ghstack-poisoned]

howardzhang-cv commented:
@pytorchbot label "topic: flex_attention" "release_notes: nn"

drisspg reviewed (commented):
Looks good, can we also add some sanity numeric checks to the test_flex_attention which use SQNR and test against this. 

I think the only way people would/should actually use this is to write a score mod that 

does something like 
```
def score_mod(score, b, h, ,q, kv):
     return  score * kv_dequant_scale
```
for instance if doing this iwth per tenosr scaling, we shoudl probably also test w/ per head scaling. tests should use fp8 and `simulate` fp8 kv cache 

Also what happens when someone tries to call bwds, is the error comprehensible?

howardzhang-cv committed:
Update on "[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU"


Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Chillee drisspg yanboliang BoyuanFeng

[ghstack-poisoned]

howardzhang-cv committed:
Update

[ghstack-poisoned]

howardzhang-cv committed:
Update

[ghstack-poisoned]

howardzhang-cv committed:
Update on "[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU"


Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Chillee drisspg yanboliang BoyuanFeng

[ghstack-poisoned]

howardzhang-cv committed:
Update

[ghstack-poisoned]

drisspg reviewed (approved):
Looks great, some small nits/comments

howardzhang-cv committed:
Update on "[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU"


Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Chillee drisspg yanboliang BoyuanFeng

[ghstack-poisoned]

howardzhang-cv commented on test/inductor/test_flex_attention.py:
```diff
@@ -2752,14 +2753,173 @@ def njt_score_mod(qk, b, h, q, kv):
+            (1, 1, 1024, 64), dtype=torch.bfloat16, device=device
+        )
+
+        key_scale = torch.max(torch.abs(key_ref)) / torch.finfo(torch.float8_e4m3fn).max
```
Looked into it for a long while and to be honest, not quite sure why the written version is better than _quantize_tensorwise? From what I gathered, _quantize_tensorwise calculates the scale in fp32, while I calculate the scale in the q.dtype (which is bf16). _quantize_tensorwise also does clamping to ensure everything stays in range (but none of the values in make_tensor get close to the edge of the range). I checked and, as expected, _quantize_tensorwise tends to get closer to the real value. However, after going through the full sdpa attention, my version where scale is calculated in bf16 always gets closer to reference output. Not quite sure why.

howardzhang-cv committed:
Update on "[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU"


Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Chillee drisspg yanboliang BoyuanFeng

[ghstack-poisoned]

howardzhang-cv committed:
Update

[ghstack-poisoned]

howardzhang-cv committed:
Update on "[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU"


Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Chillee drisspg yanboliang BoyuanFeng

Differential Revision: [D89820697](https://our.internmc.facebook.com/intern/diff/D89820697)

[ghstack-poisoned]

howardzhang-cv committed:
Update on "[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU"


Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Chillee drisspg yanboliang BoyuanFeng

Differential Revision: [D89820697](https://our.internmc.facebook.com/intern/diff/D89820697)

[ghstack-poisoned]

howardzhang-cv committed:
Update on "[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU"


Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Chillee drisspg yanboliang BoyuanFeng

Differential Revision: [D89899238](https://our.internmc.facebook.com/intern/diff/D89899238)

[ghstack-poisoned]

howardzhang-cv commented:
@howardzhang-cv has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D89899238).

howardzhang-cv commented:
@howardzhang-cv has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D89899238).

howardzhang-cv committed:
Update on "[flex_attention] adds support for low precision K/V inputs in compiled mode with GPU"


Allow Keys and Values to be in lower precision than Queries for memory efficiency.
K and V are automatically upcast to Q's dtype in the kernel before matmul operations.
In eager mode or compiled mode with CPU, error will be thrown with mixed dtypes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Chillee drisspg yanboliang BoyuanFeng

Differential Revision: [D89899238](https://our.internmc.facebook.com/intern/diff/D89899238)

[ghstack-poisoned]
